### PR TITLE
[inductor] fix avg_pool2d accuracy problem in lowering

### LINKED ISSUE #100987: [torch.compile] returns inconsistent value `AvgPool` when `ceil_mode = True`
### 🐛 Describe the bug

`torch.compile` returns inconsistent value `AvgPool` when `ceil_mode = True`. I think this issue should be similar with #94775, which is `MaxPool`

```py
import torch
import torch.nn.functional as F

torch.manual_seed(420)

x = torch.randn(1, 3, 6, 6)

class Model(torch.nn.Module):
    def forward(self, x):
        y = F.avg_pool2d(x, kernel_size=3, stride=2, padding=1, ceil_mode=True)
        return y

func = Model().to('cpu')

jit_func = torch.compile(func)

res1 = func(x) # without jit
print(res1)
# tensor([[[[ 0.0772,  0.4163,  0.3751,  0.1080],
#           [ 0.0030,  0.3709,  0.3102, -0.0654],
#           [-0.2650, -0.0227, -0.1824, -0.3878],
#           [-0.1654,  0.1697, -0.1766, -0.2419]],
# 
#          [[-0.0822, -0.1538,  0.1087,  0.2112],
#           [ 0.3352,  0.1204, -0.1107,  0.2392],
#           [ 0.1832, -0.0405, -0.3160,  0.0071],
#           [ 0.2121,  0.2613,  0.1620,  0.1483]],
# 
#          [[-0.2359, -0.5062, -0.4587, -0.2245],
#           [-0.0929, -0.1154, -0.0436, -0.1164],
#           [ 0.1771,  0.1738,  0.1792,  0.2010],
#           [ 0.1106, -0.0964, -0.3565, -0.2361]]]])

res2 = jit_func(x)
print(res2)
# tensor([[[[ 0.0772,  0.4163,  0.3751,  0.0720],
#           [ 0.0030,  0.3709,  0.3102, -0.0436],
#           [-0.2650, -0.0227, -0.1824, -0.2585],
#           [-0.1103,  0.1131, -0.1177, -0.1075]],
# 
#          [[-0.0822, -0.1538,  0.1087,  0.1408],
#           [ 0.3352,  0.1204, -0.1107,  0.1595],
#           [ 0.1832, -0.0405, -0.3160,  0.0047],
#           [ 0.1414,  0.1742,  0.1080,  0.0659]],
# 
#          [[-0.2359, -0.5062, -0.4587, -0.1497],
#           [-0.0929, -0.1154, -0.0436, -0.0776],
#           [ 0.1771,  0.1738,  0.1792,  0.1340],
#           [ 0.0738, -0.0643, -0.2377, -0.1049]]]])
```

### Versions

<details>
<summary>Click to expand</summary>

```
Collecting environment information...
PyTorch version: 2.1.0.dev20230503+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.1 LTS (x86_64)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
Clang version: 14.0.0-1ubuntu1
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.9.16 (main, Mar  8 2023, 14:00:05)  [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.19.5-051905-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.5.119
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3060
Nvidia driver version: 510.108.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   46 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          24
On-line CPU(s) list:             0-23
Vendor ID:                       GenuineIntel
Model name:                      12th Gen Intel(R) Core(TM) i9-12900K
CPU family:                      6
Model:                           151
Thread(s) per core:              2
Core(s) per socket:              16
Socket(s):                       1
Stepping:                        2
CPU max MHz:                     6700.0000
CPU min MHz:                     800.0000
BogoMIPS:                        6374.40
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault cat_l2 invpcid_single cdp_l2 ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdt_a rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi umip pku ospke waitpkg gfni vaes vpclmulqdq tme rdpid movdiri movdir64b fsrm md_clear serialize pconfig arch_lbr ibt flush_l1d arch_capabilities
Virtualization:                  VT-x
L1d cache:                       640 KiB (16 instances)
L1i cache:                       768 KiB (16 instances)
L2 cache:                        14 MiB (10 instances)
L3 cache:                        30 MiB (1 instance)
NUMA node(s):                    1
NUMA node0 CPU(s):               0-23
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Retbleed:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.24.2
[pip3] pytorch-triton==2.1.0+7d1a95b046
[pip3] torch==2.1.0.dev20230503+cu118
[pip3] torchaudio==2.1.0.dev20230503+cu118
[pip3] torchvision==0.16.0.dev20230503+cu118
[conda] numpy                     1.24.2                   pypi_0    pypi
[conda] pytorch-triton            2.1.0+7d1a95b046          pypi_0    pypi
[conda] torch                     2.1.0.dev20230503+cu118          pypi_0    pypi
[conda] torchaudio                2.1.0.dev20230503+cu118          pypi_0    pypi
[conda] torchvision               0.16.0.dev20230503+cu118          pypi_0    pypi
```

</details>

cc @ezyang @gchanan @zou3519 @soumith @msaroufim @wconstab @ngimel @bdhirsh @anijain2305

### PR DESCRIPTION
**Pull request by Valentine233**
Fixes #100987

In the current `avg_pool2d` lowering of inductor when `count_include_pad`, the mean of each window is calculated by dividing a fixing value, i.e. `kernel_size[0] * kernel_size[1]`. However for ceil mode, the amount of number in a window on the border could be less than `kernel_size[0] * kernel_size[1]`. This PR fixes the issue.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/_inductor/lowering.py [hunk 1]
```diff
@@ -2806,39 +2806,49 @@ def offset_fn(index):
...
         mask = ops.and_(
-            range_mask(ih, h),
-            range_mask(iw, w),
+            range_mask(ih, h + padding_h, -padding_h),
+            range_mask(iw, w + padding_w, -padding_w),
+        )
+        return (
+            ops.masked(
+                mask,
+                lambda: constant_boundary_condition_2d(x, 1.0)([*prefix, ih, iw]),
```
Comment (jgong5) [inline torch/_inductor/lowering.py:2846]: I guess you should pass in the padded value instead of hard-coding it as `1.0` here?

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: torch/_inductor/lowering.py (hunk 1 above)
Comment (Valentine233) [reply torch/_inductor/lowering.py]: Good point. I have fixed it.

File: torch/_inductor/lowering.py [hunk 2]
```diff
@@ -2806,39 +2806,51 @@ def offset_fn(index):
...
     return ops.and_(
-        range_mask_low(i),
-        range_mask_high(i, length),
+        range_mask_low(i, low),
+        range_mask_high(i, high),
     )
 
 
-def constant_boundary_condition_2d(x, fill_value, padding):
+def constant_boundary_condition_2d(x, fill_value, padding=None, pad_fill_value=1.0):
```
Comment (jgong5) [inline torch/_inductor/lowering.py:2830]: not sure if having `pad_fill_value` as `1.0` by default is good, or maybe use `0.0` is better?

**Review by jgong5 — APPROVED** [review]

File: torch/_inductor/lowering.py (hunk 2 above)
Comment (Valentine233) [reply torch/_inductor/lowering.py:2830]: I suppose that `pad_fill_value` is `1.0` in most cases. In fact, `pad_fill_value=0.0` indicates that there is no padding or `count_include_pad=False`, which has the same effect as `padding=None`.

File: torch/_inductor/lowering.py [hunk 3]
```diff
@@ -2806,39 +2806,51 @@ def offset_fn(index):
...
+def range_mask_high(i: sympy.Expr, high: sympy.Expr):
     return ops.lt(
         ops.index_expr(i, torch.int64),
-        ops.index_expr(length, torch.int64),
+        ops.index_expr(high, torch.int64),
     )
 
 
-def range_mask(i: sympy.Expr, length: sympy.Expr):
+def range_mask(i: sympy.Expr, high: sympy.Expr, low: sympy.Expr):
```
Comment (jansel) [inline torch/_inductor/lowering.py:2823]: low, high would make more sense as an arg order than high, low

**Review by jansel — APPROVED** [review]

**Review by EikanWang — APPROVED** [review]

File: torch/_inductor/lowering.py (hunk 3 above)
Comment (Valentine233) [reply torch/_inductor/lowering.py:2823]: Good point and I have considered this problem. I finally choose high, low instead of low, high because high is mandatory and low is optional.
